Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs

```diff
@@ -0,0 +1,273 @@
+                environmentYmlPath,
+                requirementsTxtPath,
+                ct
+            );
+        }
+
+        public static async Task ShowAddVirtualEnvironmentDialogAsync(
+            IServiceProvider site,
+            PythonProjectNode project,
+            string existingCondaEnvName = null,
+            string environmentYmlPath = null,
+            string requirementsTxtPath = null,
+            CancellationToken ct = default(CancellationToken)
+        ) {
+            await ShowDialogAsync(
+                PageKind.VirtualEnvironment,
+                site,
+                project,
+                existingCondaEnvName,
+                environmentYmlPath,
+                requirementsTxtPath,
+                ct
+            );
+        }
+
+        public static async Task ShowAddCondaEnvironmentDialogAsync(
+            IServiceProvider site,
+            PythonProjectNode project,
+            string existingCondaEnvName = null,
+            string environmentYmlPath = null,
+            string requirementsTxtPath = null,
+            CancellationToken ct = default(CancellationToken)
+        ) {
+            await ShowDialogAsync(
+                PageKind.CondaEnvironment,
+                site,
+                project,
+                existingCondaEnvName,
+                environmentYmlPath,
+                requirementsTxtPath,
+                ct
+            );
+        }
+
+        public static async Task ShowAddExistingEnvironmentDialogAsync(
+            IServiceProvider site,
+            PythonProjectNode project,
+            string existingCondaEnvName = null,
+            string environmentYmlPath = null,
+            string requirementsTxtPath = null,
+            CancellationToken ct = default(CancellationToken)
+        ) {
+            await ShowDialogAsync(
+                PageKind.ExistingEnvironment,
+                site,
+                project,
+                existingCondaEnvName,
+                environmentYmlPath,
+                requirementsTxtPath,
+                ct
+            );
+        }
+
+        private static async Task ShowDialogAsync(
+            PageKind activePage,
+            IServiceProvider site,
+            PythonProjectNode project,
+            string existingCondaEnvName = null,
+            string environmentYmlPath = null,
+            string requirementsTxtPath = null,
+            CancellationToken ct = default(CancellationToken)
+        ) {
+            if (site == null) {
+                throw new ArgumentNullException(nameof(site));
+            }
+
+            var sln = (IVsSolution)site.GetService(typeof(SVsSolution));
+            var projects = sln?.EnumerateLoadedPythonProjects().ToArray();
+
+            var projectViews = projects
+                .Select((projectNode) => new ProjectView(projectNode))
+                .ToArray();
+
+            var selectedProjectView = projectViews.SingleOrDefault(pv => pv.Node == project);
```

Comment:
We should handle the `InvalidOperationException` from here. (It'll happen when we least expect it...)